system/io.h

```diff
@@ -145,4 +147,61 @@ outb_p (unsigned char __value, unsigned short int __port)
     );
 }
 
+#elif defined(__loongarch_lp64)
+
+#define LOONGSON_IO_PROT_BASE 0x0efdfc000000
+
+#define __IN(s, s1) \
+static inline RETURN_TYPE __in##s(unsigned short port) { \
+    RETURN_TYPE _v; \
+    __asm__ __volatile__ ( \
+        "li.d $t0, 0x28\n\t" \
+        "csrwr $t0, 0x0\n\t" \
+        "ld." s1 " %0, %1, 0 \n\t" \
+        "csrwr $t0, 0x0\n\t" \
```

Comment:
Why write `0x28` to `CSR.CRMD` for port MMIO? Just issuing accesses to the mapped "port" address should be enough.

The original CRMD content is not preserved so I highly suspect this is not the correct thing to do. Do you mean some kind of barrier here? If so `dbar 0` could be appropriate I guess.

This also applies to the `out` definitions below.

Reply:
The old value is stored in $t0, please refer to Section 4.2.1 of the LoongAch Reference Manual Volume 1:
![image](https://github.com/user-attachments/assets/3ea1cca9-20cd-44e7-967e-2d7450db1d3b)

Updated `CRMD` to change the CPU to instruction fetch cacheable, load/store uncacheable, DA mode when access to an IO region is required.

Reply:
> The old value is stored in $t0, please refer to Section 4.2.1 of the LoongAch Reference Manual Volume 1: ![image](https://private-user-images.githubusercontent.com/11434289/357313447-3ea1cca9-20cd-44e7-967e-2d7450db1d3b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1MzU4MDAsIm5iZiI6MTcyMzUzNTUwMCwicGF0aCI6Ii8xMTQzNDI4OS8zNTczMTM0NDctM2VhMWNjYTktMjBjZC00NGU3LTk2N2UtMmQ3NDUwZGIxZDNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEzVDA3NTE0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzZTMxMjE4NWIzZWYzMjI4YjU0ZTA2NjcyZmQzZTJiZjQ2MmRlOTgyYzlhMzUyZTdmNzdlMjM4MDliMjA0ZmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NtHuV8nKnPa5wju4vhyOCSwde57nrh3I9W0gwQvAW6Y)
> 
> Updated `CRMD` to change the CPU to instruction fetch cacheable, load/store uncacheable, DA mode when access to an IO region is required.

Ha, thanks! Apparently I've forgotten this somehow. It's totally fine then.